snuba/api.py

```diff
@@ -263,8 +264,19 @@ def json_default(obj):
     )
 
 
-@split_query
-def parse_and_run_query(dataset, request: Request, timer):
+def split_query_by_dataset(query_func: SplitQueryFunc):
+    """
+    Splits the query based on the dataset
+    """
+
+    def wrapper(dataset, *args, **kwargs):
+        return dataset.split_query(query_func, *args, **kwargs)
```

Comment:
This approach seems a bit convoluted to me:
we have one and only one function in Snuba that performs a query given a request (parse_and_run_query), now we add a wrapper to it to perform the split, and we pass that function to a method in the dataset that decides how many times to call it and how to merge the results.
It seems to me we are doing this to go around the fact that we want (for good reasons) to make the query execution api.py depend on dataset and not the other way around. Also this avoids circular dependencies. It seems that passing this function down is kind of cheating on this dependency structure. We generally pass a closure when the logic that receives a closure is supposed to be generic with respect to the closure, which is not the case here since there is only one possible closure.
Does this make sense ?

What I had in mind tome ago when defining the new dataset structure was to represent split as query processing:
examples:
**Time split**
This is like a SQL UNION of multiple queries:
the processor takes a query object, it splits into multiple Query objects and also provides the logic to compose the results back. Still api.py is the one that runs the queries. The dataset only provides the processor class.

**col split**
It is a similar thing, the processor breaks the query object apart removing most fields, returns the new query, then receives the results back and produces the new query.

I did not put much more thoughts in the abstraction there because it was planned for much later when we will have a better query class, seems we should start thinking about that now.
Any thoughts?